tests/ManageCourses.Tests/ImporterTests/UcasDataMigratorTests.cs

```diff
@@ -106,120 +102,541 @@ public void ErroneousCourseLeavesOtherCoursesAlone()
+            updatedProvider.Courses.Count.Should().Be(1);
+            updatedProvider.Courses.Single(c => c.CourseCode == courseCode)
+                .AccreditingProvider.ProviderCode.Should().Be(accreditingProviderCode2);
+        }
+
+        [Test]
+        public void CourseOptedIn_AccreditingProviderOptedIn()
+        {
+            // note: we can't test an import of a new provider's course where they are opted in because
+            // they have to be in our database already to be opted in
+
+            const string instCode = "AA1";
+            const string accreditingProviderCode = "AINST201";
+            const string accreditingProviderCode2 = "AINST202";
+            const string courseCode = "CRS101";
+            const string unmodifiedCourseName = "Primary";
+            const string campusCode = "CAMP101";
+
+            // arrange
+            Context.Providers.Add(new ProviderBuilder()
+                .WithCode(instCode)
+                .WithOptedIn()
+                .WithCourses(new List<Course>
+                {
+                    new CourseBuilder()
+                        .WithCode(courseCode)
+                        .WithName(unmodifiedCourseName)
+                        .WithAccreditingProvider(new ProviderBuilder()
+                            .WithOptedIn()
+                            .WithCode(accreditingProviderCode))
+                }));
+            Context.Providers.Add(new ProviderBuilder()
+                .WithOptedIn()
+                .WithCode(accreditingProviderCode2));
+            Context.Save();
+
+            UcasCourse ucasCourse = new PayloadCourseBuilder()
+                .WithInstCode(instCode)
+                .WithCrseCode(courseCode)
+                .WithAccreditingProvider(accreditingProviderCode2)
+                .WithCampusCode(campusCode);
+            var ucasPayload = new PayloadBuilder()
+                .WithInstitutions(new List<UcasInstitution>
+                {
+                    new PayloadInstitutionBuilder()
+                        .WithInstCode(instCode)
+                })
+                .WithCourses(new List<UcasCourse> {ucasCourse})
+                .WithCampuses(new List<UcasCampus>
+                {
+                    new PayloadCampusBuilder()
+                        .WithCampusCode(campusCode)
+                        .WithInstCode(instCode)
+                });
+
+            // act
+            DoImport(ucasPayload);
+
+            // assert
+            var provider = Context.Providers.Single(p => p.ProviderCode == instCode);
+            provider.Courses.Count.Should().Be(1);
+            provider.Courses.Single(c => c.CourseCode == courseCode)
+                .AccreditingProvider.ProviderCode.Should().Be(accreditingProviderCode);
+        }
+
+        [Test]
+        public void AllVariations()
+        {
+            // build a big payload with all of the above
```

Comment:
xdoc instead of of trailing comments here and there and somewhere, you forcing others to read titbits to understand whats what.